File: backend/app/utils/categories.py
DEFAULT_CATEGORIES = [
    "Flight",
    "Accommodation",
    "Food & Drink",
    "Transport",
    "Tour & Activities",
    "Entertainment",
    "Shopping",
    "Gift",
    "Health & Medical",
    "Communication",
    "Other",
]

CATEGORY_EMOJI = {
    "Flight": "✈️",
    "Accommodation": "🏨",
    "Food & Drink": "🍽️",
    "Transport": "🚗",
    "Tour & Activities": "🎟️",
    "Entertainment": "🎉",
    "Shopping": "🛍️",
    "Gift": "🎁",
    "Health & Medical": "💊",
    "Communication": "📱",
    "Other": "📌",
}


def validate_category(category: str, custom_categories: list[str]) -> bool:
    return category in DEFAULT_CATEGORIES + custom_categories


def get_emoji(category: str, custom_emoji_map: dict[str, str] | None = None) -> str:
    if custom_emoji_map and category in custom_emoji_map:
        return custom_emoji_map[category]
    return CATEGORY_EMOJI.get(category, "🏷️")

File: android/app/src/main/java/com/duitrip/app/domain/Categories.kt
package com.duitrip.app.domain

/** Ported 1:1 from backend/app/utils/categories.py and frontend DEFAULT_CATEGORIES. */
object Categories {

    data class DefaultCategory(val name: String, val emoji: String)

    val DEFAULT_CATEGORIES: List<DefaultCategory> = listOf(
        DefaultCategory("Flight", "✈️"),
        DefaultCategory("Accommodation", "🏨"),
        DefaultCategory("Food & Drink", "🍽️"),
        DefaultCategory("Transport", "🚗"),
        DefaultCategory("Tour & Activities", "🎟️"),
        DefaultCategory("Entertainment", "🎉"),
        DefaultCategory("Shopping", "🛍️"),
        DefaultCategory("Gift", "🎁"),
        DefaultCategory("Health & Medical", "💊"),
        DefaultCategory("Communication", "📱"),
        DefaultCategory("Other", "📌"),
    )

    private val EMOJI_BY_NAME: Map<String, String> =
        DEFAULT_CATEGORIES.associate { it.name to it.emoji }

    val DEFAULT_NAMES: List<String> = DEFAULT_CATEGORIES.map { it.name }

    /** Max custom categories per trip — matches the old backend invariant. */
    const val MAX_CUSTOM_CATEGORIES = 20

    fun isValid(category: String, customNames: List<String>): Boolean =
        category in DEFAULT_NAMES || category in customNames

    fun emojiFor(category: String, customEmojiMap: Map<String, String>? = null): String {
        customEmojiMap?.get(category)?.let { return it }
        return EMOJI_BY_NAME[category] ?: "🏷️"
    }
}
